# Bit Manipulation


## Binary Operators
```
&    binary and
|    binary or
^    binary xor (exclusive or)
~    binary ones complement effect of  flipping bits
<<   binary left shift
>>  binary right shift

    

x  y    x^y
0  0    0    
0  1    1
1  0    1
1  1    0   
    
```
https://www.hackerearth.com/practice/basic-programming/bit-manipulation/basics-of-bit-manipulation/tutorial/
    
https://www.youtube.com/watch?v=F0RywIXmqlU&list=PL5tcWHG-UPH1u5iox6v1Hey59vNd5cnTw

## Printing Binary Numbers Formating binary

In [239]:
i = 3
print(f'{i:>2} in binary is {i:>08b}')
print(f'{i:>2} in binary is {i:<08b}')
print("{0:b} | {0}".format(5))

# converting
print(bin(19))
print(int(0b100101))

 3 in binary is 00000011
 3 in binary is 11000000
101 | 5
0b10011
37


### Get Num to Binary

In [192]:
n = 44
print(bin(n)[2:])

101100


## Basic Operations

x << y  
Returns x with the bits shifted to the left by y places (and new bits on the right-hand-side are zeros). This is the same as multiplying x by 2**y.

x >> y  
Returns x with the bits shifted to the right by y places. This is the same as //'ing x by 2**y.

x & y  
Does a "bitwise and". Each bit of the output is 1 if the corresponding bit of x AND of y is 1, otherwise it's 0.

x | y  
Does a "bitwise or". Each bit of the output is 0 if the corresponding bit of x AND of y is 0, otherwise it's 1.

~ x  
Returns the complement of x - the number you get by switching each 1 for a 0 and each 0 for a 1. This is the same as -x - 1.

x ^ y  
Does a "bitwise exclusive or". Each bit of the output is the same as the corresponding bit in x if that bit in y is 0, and it's the complement of the bit in x if that bit in y is 1.




```
a =   0011 1100
b =   0000 1101
-----------------
a&b = 0000 1100    and
a|b = 0011 1101     or
a^b = 0011 0001     xor  
~a  = 1100 0011     not or complement
```

In [160]:


a = 7          # 60 = 0011 1100 
b = 4           # 13 = 0000 1101 
c = 0

c = a & b;        # 12 = 0000 1100
print("{0} : {0:b}, {1} : {1:b}  {0} & {1} = {2:b} | {2}".format(a, b, c))

c = a | b;        # 61 = 0011 1101 
print("{0} : {0:b},  {1} : {1:b}  {0} | {1} = {2:b} | {2}".format(a, b, c))

c = a ^ b;        # 49 = 0011 0001
print("{0} : {0:b},  {1} : {1:b}  {0} ^ {1} = {2:b} | {2}".format(a, b, c))

c = ~a;           # -61 = 1100 0011
print("{0} : {0:b},  ~{0}  = {1:b} | {1}".format(a, c ))

1
7 : 111, 1 : 1  7 & 1 = 1
True
False
7 : 111, 4 : 100  7 & 4 = 100 | 4
7 : 111,  4 : 100  7 | 4 = 111 | 7
7 : 111,  4 : 100  7 ^ 4 = 11 | 3
7 : 111,  ~7  = -1000 | -8


## Bit Facts
0s and 1s is a sequence of the same bits

In [139]:
x = 9
print(f'8 in binary {x:b}')

# x ^ 0s = x
print(f'x ^ 0s = x {x^0b0000:b}')

# x ^ 1s = ~x
print(f'x ^ 1s = ~x {x^0b1111:>04b}')

# x ^ x = 0
print(f'x ^ x  = 0 {x^x:>04b}')

# x & 0s = 0
print(f'x & 0s = 0 {x&0b0000:>04b}')

# x & 1s = x
print(f'x & 1s = x {x&0b1111:>04b}')

# x & x = x
print(f'x & x  = x {x&x:>04b}')

# x | 0s = x
print(f'x | 0s  = x {x|0b0000:>04b}')

# x | 1s = 1s
print(f'x | 1s  = x {x|0b1111:>04b}')

# x | x = x
print(f'x | x  = x {x|x:>04b}')

8 in binary 1001
x ^ 0s = x 1001
x ^ 1s = ~x 0110
x ^ x  = 0 0000
x & 0s = 0 0000
x & 1s = x 1001
x & x  = x 1001
x | 0s  = x 1001
x | 1s  = x 1111
x | x  = x 1001


## Bit Shifting

Left Shift  
<< inserts zeros at the right end of the bit, so bits are shifted left   
    x << y is multiplying  x by 2\*\*y.   
    
Right Shift   
\>> inserts zeros at the left end of the bit, so bits are shifted right   
    x \>> y is dividing x by 2\*\*y  

multiplying or dividing a number by powers of  two

```
n << 1  -  n * 2
n << 2  -  n * 4

```

In [180]:
a = 10         
b = 4          

# shift 1 to left
c = a << 1;       
print("{0} : bin {0:b},  {0} << 1 = {1:b} | {1}".format(a, c))

# shift 2 to left
c = a << 2;       
print("{0} : bin {0:b},  {0} << 2 = {1:b} | {1}".format(a, c))

a = 6
# shift 1 to right
c = a >> 1;      
print("{0} : bin {0:b},  {0} >> 1 = {1:b} | {1}".format(a, c ))    

a = 20
# shift 2 to right
c = a >> 2;       
print("{0} : bin {0:b},  {0} >> 2 = {1:b} | {1}".format(a, c ))    

10 : bin 1010,  10 << 1 = 10100 | 20
10 : bin 1010,  10 << 2 = 101000 | 40
6 : bin 110,  6 >> 1 = 11 | 3
20 : bin 10100,  20 >> 2 = 101 | 5


<hr>

## Basic Usage



### Check if a bit is set

In [238]:
def bit_set(position, binary):
    # create bit mask and 
    bit_mask = 1 << position
    print(f'bit mask {bit_mask:b}')
    # use & operator
    return binary & bit_mask == bit_mask

binary = 12
position = 1 

print("{0} : {0:b}, mask : {1:>08b}  {0:b} & {1:>08b} == {1:>08b} is {2}".format(binary, position, bit_set(position, binary)))

# & how to check if a bit is checked  - 
a = 7
b = 1
"""
    is 001  in 111 - use the and operator
"""
print(a & b)
print("7 : {0:b}, mask : {1:b}  7 & 1 = {2}".format(a, b, a & b))
print(a & b == b)
print(a+1 & b == b)

bit mask 10
12 : 1100, mask : 00000001  1100 & 00000001 == 00000001 is False
1
7 : 111, mask : 1  7 & 1 = 1
True
False



### Set bit position
binary in this example is zero

In [79]:
def set_bit(position, binary):   
    # set up mask
    bit_mask = 1 << position
    # use or operator to set the bit
    return bit_mask | binary

print('set this binary {1:>08b} position {2} dec {0} binary {0:b} '.format(set_bit(4, 0b000000000), 0b000000000, 4))

print('set this binary {1:>08b} position {2} dec {0} binary {0:b} '.format(set_bit(2, 0b000010000), 0b000010000, 2))

set this binary 00000000 position 4 dec 16 binary 10000 
set this binary 00010000 position 2 dec 20 binary 10100 


### Get Bit

In [218]:
def get_bit(position, binary):
    # create bit mask and 
    bit_mask = 1 << position
    # use & operator
    return binary & bit_mask != 0

print('found bit {0} binary bit found {0:b} '.format(get_bit(1, 0b000011101)))
print('found bit {0} binary  bit found {0:b} '.format(get_bit(0, 0b000011101)))  
print('found bit {0} binary   bit found {0:b} '.format(get_bit(4, 0b000011101)))  

found bit False binary bit found 0 
found bit True binary  bit found 1 
found bit True binary   bit found 1 


### Clear Bit

In [94]:
def clear_bit(pos, num):
    mask = ~(1 << pos)
    print(bin(mask))
    return num & mask

print('dec {0} before binary {1:b} after clearing {0:b} '.format(clear_bit(0, 0b000011101), 0b000011101))

-0b10
dec 28 before binary 11101 after clearing 11100 


### Update Bit

In [152]:
def update_bit(num, i, bit_is_1):
    value = 1 if bit_is_1 else 0
    mask = ~(1 << i)
    return (num & mask) | (value << i)

print(f'7 binary {7:b} {update_bit(7, 1, False):b}')


7 binary 111 101


### Bitwise way to switch between true and false

In [141]:
result = 0
for _ in range(10):
    print(result)
    result ^= 1    

0
1
0
1
0
1
0
1
0
1


### Bitwise return odd or even of number

x & 1  
0 = even  
1 = odd  


In [176]:
for i in range(32):
    print(f'{i} & 1 = {i & 1}')

0 & 1 = 0
1 & 1 = 1
2 & 1 = 0
3 & 1 = 1
4 & 1 = 0
5 & 1 = 1
6 & 1 = 0
7 & 1 = 1
8 & 1 = 0
9 & 1 = 1
10 & 1 = 0
11 & 1 = 1
12 & 1 = 0
13 & 1 = 1
14 & 1 = 0
15 & 1 = 1
16 & 1 = 0
17 & 1 = 1
18 & 1 = 0
19 & 1 = 1
20 & 1 = 0
21 & 1 = 1
22 & 1 = 0
23 & 1 = 1
24 & 1 = 0
25 & 1 = 1
26 & 1 = 0
27 & 1 = 1
28 & 1 = 0
29 & 1 = 1
30 & 1 = 0
31 & 1 = 1


### Get rightmost lowest bit 1 in binary - two ways

In [194]:
x = 10  # 1010
print(f'{x ^ (x & (x - 1)):>08b}')


x = 10  # 1010
print(f'{x & ~(x - 1):>08b}')

00000010
00000010


### Clear Right Most Bit
x & (x-1)

In [108]:
x = 10
print(f'{x:b}')
x = x&(x-1)
print(f'{x:b}')

1010
1000


## Coding Problems

### check parity of word 
even 1's it is 0  
odd 1's it is 1

In [103]:
## Brute Force O(n)
def parity(x):
    result = 0
    while x:
        result ^= x & 1
        x >>= 1
    return result

print(f'11 {11:b} {parity(11)}')
print(f'10 {10:b} {parity(10)}')

11 1011 1
10 1010 0


In [114]:
# drop right most bit o(k) k number of bit set to 1
def parity(x):
    result = 0
    while x:  # checks for x to be zero
        result ^= 1 #bitwise switch between true and false
        print(f'{result:b}')
        print(f'{x:b}')
        x &= x - 1  # remove each left 1
        
    return result

print(f'11 {11:b} {parity(11)}')
print(f'10 {10:b} {parity(10)}')

1
1011
0
1010
1
1000
11 1011 1
1
1010
0
1000
10 1010 0


### Count all bits set to 1 in a positive integer

test bits one at a time from LSB left to right. uses shifting and and masking

In [68]:
def count_bits(x):
    num_bits = 0
    while x:
        num_bits += x & 1
        x >>= 1
    return num_bits

print(f'count bits: {count_bits(8)} bits {8:b}')
print(f'count bits: {count_bits(7)} bits {7:b}')

count bits: 1 bits 1000
count bits: 3 bits 111


### Find Element that Appears Once In A list
Given an array of integers where every integer appears three times except for one integer, which only occurs once, find and return it.

In [188]:
def find_unique_threes(arr):
    result_arr = [0] * 32
    
    for num in arr:
        for i in range(32):
            # print(f'num : {num}   {num} >>  {i}  = {num >> i}')
            
            bit = (num >> i) & 1
            
            # print(f'bit : {bit:b}')
            
            result_arr[i] += bit
            
           #  print(result_arr)
    result = 0
    for i, bit in enumerate(result_arr):
        if bit % 3 != 0:
            result += 2 ** i
    
    return result
            
arr = [6,1,3,3,3,6,6]

print(find_unique_threes(arr))

            
arr = [13,19,13,13]

print(find_unique_threes(arr))

1
19


### Compute the longest consecutive string of ones in binary
10011100 would return 3  

n & (n << 1)      
trick to remove consecutive strings of 1's one at a time.   
```
10110  
shift it  
 101100  
& 10110  
 ------  
 000100 only one 1
```


In [206]:
def find_length(n):
    max_length = 0
    
    while n:
        max_length += 1
        print(f'n << 1 = {n << 1}')
        print(f'{n} & ({n << 1}) = {n & (n << 1)} {n & (n << 1):b}')
        print(f'{n:b} & {n << 1:b}')
        
        n = n & (n << 1)
    return max_length
    
find_length(21)

n << 1 = 42
21 & (42) = 0 0
10101 & 101010


1

### Find nth sevenish number
1, 7, 8, 49
power of 7 or sum of unique powers of seven

In [214]:
def find_seven(n):
    answer = 0
    bit_place = 0
    
    while n:
        if (n & 1):  # odd even check
            answer += 7 ** bit_place  # just math no bitwise operatins
        print(f'n {n:b}')
        
        n >>= 1 
        bit_place += 1
        
        print(f'{answer:b}')
    
    return answer

print(find_seven(2))

n 10
0
n 1
111
7


## XOR  Exclusive 

```
x  y    x^y
0  0    0    
0  1    1
1  0    1
1  1    0   
    
```

XOR is a logical bitwise operator that returns 0 (false) if both bits are the same and returns 1 (true) otherwise. In other words, it only returns 1 if exactly one bit is set to 1 out of the two bits in comparison.


Important properties of XOR to remember #
Following are some important properties of XOR to remember:

1. Taking XOR of a number with itself returns 0, e.g.,

    1 ^ 1 = 0  
    29 ^ 29 = 0  
2. Taking XOR of a number with 0 returns the same number, e.g.,

    1 ^ 0 = 1  
    31 ^ 0 = 31  
    
3. XOR is Associative & Commutative, which means:

    (a ^ b) ^ c = a ^ (b ^ c)  
    a ^ b = b ^ a   

In [27]:
a = 2          
b = 4          
c = 0

c = a ^ b     
print("{0} : {0:>08b},  {1} : {1:>08b}  {0} ^ {1} = {2:>08b} | {2}".format(a, b, c))

2 : 00000010,  4 : 00000100  2 ^ 4 = 00000110 | 6


### Comparison of 2 Boolean Functions x and y.

If x and y are same then x ^ y = 0  
If x and y are different then x ^ y = 1

In [22]:
x = True
y = False

print(x ^ y)

x = True
y = True

print(x ^ y)

x = False
y = False

print(x ^ y)

True
False
False


### Given an array of n-1 integers in the range from 1 to n, find the one number that is missing from the array.
Input: 1, 5, 2, 6, 4 
Answer: 3 

In [1]:
def find_missing_number(arr):
  n = len(arr) + 1
  # find sum of all numbers from 1 to n.
  s1 = 0
  for i in range (1, n+1):
    s1 += i

  # subtract all numbers in input from sum.
  for i in arr:
    s1 -= i
  
  # s1, now, is the missing number
  return s1

def main():
  arr = [1, 5, 2, 6, 4] 
  print('Missing number is:' + str(find_missing_number(arr)))

main()

Missing number is:3


Remember the important property of XOR that it returns 0 if both the bits in comparison are the same. In other words, XOR of a number with itself will always result in 0. This means that if we XOR all the numbers in the input array with all numbers from the range 11 to nn then each number in the input is going to get zeroed out except the missing number. Following are the set of steps to find the missing number using XOR:

1. XOR all the numbers from 1 to n, let’s call it x1.  
2. XOR all the numbers in the input array, let’s call it x2.  
3. The missing number can be found by x1 XOR x2.  

In [48]:
def find_missing_number(arr):
    n = len(arr) + 1
    
    # x1 represents XOR of all values from 1 to n
    x1 = 1
    for i in range(2, n+1):
        x1 = x1 ^ i
    print(f'{x1:b}')
    
    # x2 represents XOR of all values in arr
    x2 = arr[0]
    for i in range(1, n-1):
        x2 = x2 ^ arr[i]
    print(f'{x2:>b}')
    
    # missing number is the xor of x1 and x2
    return x1 ^ x2

arr = [1, 5, 2, 6, 4] 
print('Missing number is: ' + str(find_missing_number(arr)))

# 01
# 10 ^
# 11
# 11 ^
# 00

# 01
# 11^
# 10

# 00
# 10 ^
# 10 2

111
100
Missing number is: 3


' 3 in binary is 00000011'

### Single Number

In a non-empty array of integers, every number appears twice except for one, find that single number.

Input: 1, 4, 2, 1, 3, 2, 3   
Output: 4  
Example 2:  

Input: 7, 9, 7  
Output: 9  

In [50]:
def find_single_number(arr):
    # TODO: Write your code here
    x1 = arr[0]
    for i in range(1, len(arr)):
        x1 = x1 ^ arr[i]
        
    
    
    return x1

def main():
    arr = [1, 4, 2, 1, 3, 2, 3]
    print(find_single_number(arr))

main()

4


In [52]:
def find_single_number(arr):
    num = 0
    for i in arr:
        num ^= i
    return num

arr = [1, 4, 2, 1, 3, 2, 3]
print(find_single_number(arr))

4


### Two Single Numbers (medium)
In a non-empty array of numbers, every number appears exactly twice except two numbers that appear only once. Find the two numbers that appear only once.

Input: [1, 4, 2, 1, 3, 5, 6, 2, 3, 5]  
Output: [4, 6]  


Input: [2, 1, 3, 2]  
Output: [1, 3]  


This problem is quite similar to Single Number, the only difference is that, in this problem, we have two single numbers instead of one. Can we still use XOR to solve this problem?

Let’s assume num1 and num2 are the two single numbers. If we do XOR of all elements of the given array, we will be left with XOR of num1 and num2 as all other numbers will cancel each other because all of them appeared twice. Let’s call this XOR n1xn2. Now that we have XOR of num1 and num2, how can we find these two single numbers?

As we know that num1 and num2 are two different numbers, therefore, they should have at least one bit different between them. If a bit in n1xn2 is ‘1’, this means that num1 and num2 have different bits in that place, as we know that we can get ‘1’ only when we do XOR of two different bits, i.e.,

1 XOR 0 = 0 XOR 1 = 1
We can take any bit which is ‘1’ in n1xn2 and partition all numbers in the given array into two groups based on that bit. One group will have all those numbers with that bit set to ‘0’ and the other with the bit set to ‘1’. This will ensure that num1 will be in one group and num2 will be in the other. We can take XOR of all numbers in each group separately to get num1 and num2, as all other numbers in each group will cancel each other. Here are the steps of our algorithm:

1. Taking XOR of all numbers in the given array will give us XOR of num1 and num2, calling this XOR as n1xn2.
2. Find any bit which is set in n1xn2. We can take the rightmost bit which is ‘1’. Let’s call this rightmostSetBit.
3. Iterate through all numbers of the input array to partition them into two groups based on rightmostSetBit. Take XOR of all numbers in both the groups separately. Both these XORs are our required numbers.

In [58]:
def find_single_numbers(nums):
    # get the XOR of the all the numbers
    n1xn2 = 0
    for num in nums:
        n1xn2 ^= num

    # get the rightmost bit that is '1'
    rightmost_set_bit = 1
    while (rightmost_set_bit & n1xn2) == 0:
        rightmost_set_bit = rightmost_set_bit << 1
    print(f'{rightmost_set_bit:b}')
    rightmost_set_bit = n1xn2 ^ (n1xn2 & (n1xn2 - 1))
    print(f'{rightmost_set_bit:b}')
    
    num1, num2 = 0, 0
    
    

    for num in nums:
        if (num & rightmost_set_bit) != 0:  # the bit is set
            num1 ^= num
        else:  # the bit is not set
            num2 ^= num

    return [num1, num2]


def main():
    print('Single numbers are:' +
          str(find_single_numbers([1, 4, 2, 1, 3, 5, 6, 2, 3, 5])))
    print('Single numbers are:' + str(find_single_numbers([2, 1, 3, 2])))


main()


10
10
Single numbers are:[6, 4]
10
10
Single numbers are:[3, 1]


### Complement of Base 10 Number (medium)

Every non-negative integer N has a binary representation, for example, 8 can be represented as “1000” in binary and 7 as “0111” in binary.

The complement of a binary representation is the number in binary that we get when we change every 1 to a 0 and every 0 to a 1. For example, the binary complement of “1010” is “0101”.

For a given positive number N in base-10, return the complement of its binary representation as a base-10 integer.

Example 1:

Input: 8  
Output: 7  
Explanation: 8 is 1000 in binary, its complement is 0111 in binary, which is 7 in base-10.

In [62]:
def calculate_bitwise_complement(num):
    # count number of total bits in 'num'
    bit_count, n = 0, num
    while n > 0:
        bit_count += 1
        n = n >> 1

    # for a number which is a complete power of '2' i.e., it can be written as pow(2, n), if we
    # subtract '1' from such a number, we get a number which has 'n' least significant bits set to '1'.
    # For example, '4' which is a complete power of '2', and '3' (which is one less than 4) has a binary
    # representation of '11' i.e., it has '2' least significant bits set to '1'
    all_bits_set = pow(2, bit_count) - 1
    print(f'{all_bits_set:b}')
    # from the solution description: complement = number ^ all_bits_set
    return num ^ all_bits_set


print('Bitwise complement is: ' + str(calculate_bitwise_complement(8)))
print('Bitwise complement is: ' + str(calculate_bitwise_complement(10)))


1111
Bitwise complement is: 7
1111
Bitwise complement is: 5


### Given a binary matrix representing an image, we want to flip the image horizontally, then invert it.

To flip an image horizontally means that each row of the image is reversed. For example, flipping [0, 1, 1] horizontally results in [1, 1, 0].

To invert an image means that each 0 is replaced by 1, and each 1 is replaced by 0. For example, inverting [1, 1, 0] results in [0, 0, 1].

```
Input: [
  [1,0,1],
  [1,1,1],
  [0,1,1]
]
Output: [
  [0,1,0],
  [0,0,0],
  [0,0,1]
]```

Solution

Flip: We can flip the image in place by replacing ith element from left with the ith element from the right.

Invert: We can take XOR of each element with 1. If it is 1 then it will become 0 and if it is 0 then it will become 1.

In [63]:
def flip_an_invert_image(matrix):
  C = len(matrix)
  for row in matrix:
    for i in range((C+1)//2):
      row[i], row[C - i - 1] = row[C - i - 1] ^ 1, row[i] ^ 1
      
  return matrix

def main():
    print(flip_an_invert_image([[1,0,1], [1,1,1], [0,1,1]]))
    print(flip_an_invert_image([[1,1,0,0],[1,0,0,1],[0,1,1,1],[1,0,1,0]]))

main()

[[0, 1, 0], [0, 0, 0], [0, 0, 1]]
[[1, 1, 0, 0], [0, 1, 1, 0], [0, 0, 0, 1], [1, 0, 1, 0]]
